src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,173 @@
+    /// but no <see cref="DataMemberAttribute"/>.
+    /// </summary>
+    public static class RequiredValidationHelper
+    {
+        private static ConcurrentDictionary<Type, Dictionary<Type, List<string>>> cachedValidationErrors
+            = new ConcurrentDictionary<Type, Dictionary<Type, List<string>>>();
+
+        public static void Validate([NotNull] Type modelType, [NotNull] ModelStateDictionary modelStateDictionary)
+        {
+            var visitedTypes = new HashSet<Type>();
+
+            // Every node maintains a dictionary of Type => Errors. 
+            // It's a dictionary as we want to avoid adding duplicate error messages.
+            // Example:
+            // Type 'A' has a child node of type 'B' somewhere in the left sub-tree
+            // and it has type 'B' again  somewhere down the righ sub-tree. From the perspective of
+            // type 'A', it should not have duplicate error messgaes for type 'B'.
+            var rootNodeValidationErrors = new Dictionary<Type, List<string>>();
+
+            Validate(modelType, visitedTypes, rootNodeValidationErrors);
+
+            foreach (var validatoinError in rootNodeValidationErrors)
+            {
+                foreach (var validationErrorMessage in validatoinError.Value)
+                {
+                    modelStateDictionary.TryAddModelError(
+                        validatoinError.Key.FullName,
+                        validationErrorMessage);
+                }
+            }
+        }
+
+        private static void Validate(
+            Type modelType,
+            HashSet<Type> visitedTypes,
+            Dictionary<Type, List<string>> currentNodeValidationErrors)
+        {
+            if (modelType.IsGenericType())
+            {
+                var enumerableOfT = modelType.ExtractGenericInterface(typeof(IEnumerable<>));
+                if (enumerableOfT != null)
+                {
+                    modelType = enumerableOfT.GetGenericArguments()[0];
+                }
+            }
+
+            if (ExcludeTypeFromValidation(modelType))
+            {
+                return;
+            }
+
+            // Avoid infinite loop in case of self-referencing properties
+            if (visitedTypes.Contains(modelType))
+            {
+                return;
+            }
+
+            visitedTypes.Add(modelType);
+
+            Dictionary<Type, List<string>> cachedCurrentNodeValidationErrors;
+            if (cachedValidationErrors.TryGetValue(modelType, out cachedCurrentNodeValidationErrors))
+            {
+                foreach (var validationError in cachedCurrentNodeValidationErrors)
+                {
+                    currentNodeValidationErrors.Add(validationError.Key, validationError.Value);
+                }
+
+                return;
+            }
+
+            foreach (var propertyInfo in modelType.GetProperties(BindingFlags.Public | BindingFlags.Instance))
```

Comment:
Regarding comment 1: For nullable types, with the current code it would be excluded from validation on the recursive call of that type (line 111 and 63)